File: harness-engineering/src/lib/chat/stream-state.ts
interface StreamState {
  chatId: string
  frontendSessionId: string
  engine: string
  streamContent: string
  status: "running" | "completed" | "failed" | "killed"
}

const streams = new Map<string, StreamState>()

export function registerStream(chatId: string, frontendSessionId: string, engine: string): void {
  streams.set(chatId, { chatId, frontendSessionId, engine, streamContent: "", status: "running" })
}

export function appendStreamContent(chatId: string, content: string): void {
  const state = streams.get(chatId)
  if (state) state.streamContent += content
}

export function setStreamStatus(chatId: string, status: StreamState["status"]): void {
  const state = streams.get(chatId)
  if (state) state.status = status
}

export function getStream(chatId: string): StreamState | undefined {
  return streams.get(chatId)
}

export function getStreamBySessionId(frontendSessionId: string): StreamState | undefined {
  for (const state of streams.values()) {
    if (state.frontendSessionId === frontendSessionId) return state
  }
  return undefined
}

export function removeStream(chatId: string): void {
  streams.delete(chatId)
}

File: harness-engineering-py/backend/app/services/stream_state.py
from typing import Dict, Optional
from app.models.schemas import StreamState


_streams: Dict[str, StreamState] = {}


def register_stream(chat_id: str, frontend_session_id: str, engine: str) -> None:
    _streams[chat_id] = StreamState(
        chat_id=chat_id,
        frontend_session_id=frontend_session_id,
        engine=engine,
    )


def append_stream_content(chat_id: str, content: str) -> None:
    state = _streams.get(chat_id)
    if state:
        state.stream_content += content


def set_stream_status(chat_id: str, status: str) -> None:
    state = _streams.get(chat_id)
    if state:
        state.status = status


def get_stream(chat_id: str) -> Optional[StreamState]:
    return _streams.get(chat_id)


def get_stream_by_session_id(frontend_session_id: str) -> Optional[StreamState]:
    for state in _streams.values():
        if state.frontend_session_id == frontend_session_id:
            return state
    return None


def remove_stream(chat_id: str) -> None:
    _streams.pop(chat_id, None)
